src/ppx_variants_conv.ml

```diff
@@ -396,21 +506,71 @@ module Gen_str = struct
               ~lhs:(Variant_constructor.pattern_without_binding v)
               ~rhs:[%expr true]
           in
-          match_fun ~name ~true_case ~false_expr:[%expr false]
+          let false_expr = [%expr false] in
+          let cases =
+            if multiple_cases
+            then [ true_case; case ~guard:None ~lhs:[%pat? _] ~rhs:false_expr ]
+            else [ true_case ]
+          in
+          if Variant_constructor.is_gadt v then 
+              let (variant_ty,lbls) = Core_type.newtypes variant_ty in
+              let pat_arg  = ppat_constraint ~loc (ppat_var ~loc {loc;txt="t"} ) 
+                variant_ty in
+              let ident = pexp_ident ~loc  {loc;txt=Longident.Lident "t"}  in
+              let match_expr = pexp_match ~loc ident cases in
+              let fun_expr  =pexp_fun ~loc Nolabel None pat_arg match_expr in 
+              let lbl_locs = List.map lbls ~f:(fun txt -> {loc;txt}) in
+              let body = with_newtypes loc lbl_locs fun_expr in
+              (* attributes are a pain to add and mean we can't use metaquot
+                 so we repeat ourselves here rather than use [locally_abstract_match] *)
+              [%stri let [%p pvar ~loc name] =  
+                [%e body] 
+                [@@warning "-4"]
+              ]
+          else
+            [%stri let [%p pvar ~loc name] = 
+              [%e pexp_function ~loc cases] [@@warning "-4"]]
         in
+
         let getter =
           let name = uncapitalized ^ "_val" in
           let pat, expr = Variant_constructor.to_getter_case v in
           let true_case =
             case ~guard:None ~lhs:pat ~rhs:[%expr Stdlib.Option.Some [%e expr]]
           in
-          match_fun ~name ~true_case ~false_expr:[%expr Stdlib.Option.None]
+          let false_expr = [%expr Stdlib.Option.None] in
+          let cases =
+            if multiple_cases
+            then [ true_case; case ~guard:None ~lhs:[%pat? _] ~rhs:false_expr ]
+            else [ true_case ]
+          in
+          match Variant_constructor.return_ty_opt v  with
+          | Some return_ty when 
+              not @@ Variant_constructor.has_existential v ->
+            let (return_ty,lbls) = Core_type.newtypes return_ty in
+            let pat_arg  = ppat_constraint ~loc (ppat_var ~loc {loc;txt="t"} ) 
+              return_ty in
+            let ident = pexp_ident ~loc  {loc;txt=Longident.Lident "t"}  in
+            let match_expr = pexp_match ~loc ident cases in
+            let fun_expr  =pexp_fun ~loc Nolabel None pat_arg match_expr in 
+            let lbl_locs = List.map lbls ~f:(fun txt -> {loc;txt}) in
+            let body = with_newtypes loc lbl_locs fun_expr in
+            (* attributes are a pain to add and mean we can't use metaquot
+               so we repeat ourselves here rather than use [locally_abstract_match] *)
+            Some [%stri let [%p pvar ~loc name] =  
+              [%e body] 
+              [@@warning "-4-56"]
```

Comment:
Why do we need -56?

Reply:
For the test example you don't; however, if you remove all but the Int, Bool, and Pair constructors, the type of the pair_val function means we have refined to only that case. The type checker sees this and Aggie at replacing with a refutation case. The suppression is not always needed but the analysis would be a little involved to determine exactly when it is. (Apologies for formatting, on phone)

Reply:
Ok. Indeed. But in that case, the code is irregular: either the _val functions take a `t` whose parameters are constrained, or `t` that's not constrained, but it's weird that some _val functions pick one and others pick the other.

Reply:
I don't think it's possible to write a getter for these cases with a consistent signature; consider what we would have if we did write the getter for `pair_val` with a consistent signature:

```
let pair_val (type a)  (t : a t) =
  match t with
  | Pair (v0, v1) -> Stdlib.Option.Some (v0,v1)
  | _ -> Stdlib.Option.None[@@warning "-4-56"]
```

The typechecker sees we are trying to return values with existentially quantified, rigid types (`$0 t` and `$1 t`) so it complains:

```
Error: this expression has type $0 t but an expression was expected of type 'a
The constructor $0 would escape it's scope
```

I guess the choice is to either have non-regular getter for non-regular types or not to generate in these cases. What do you think?

Reply:
I don't agree that things have to be irregular.

You can either generalize

```
type _ t = A : int -> bool t

val a : 'z t -> (('z, bool) Type_equal.t * int) option
(* or *)
val a : bool t -> int option
```

or skip the getter.

What's up your use case? I'm ok with not providing a function until someone cares about them. Otherwise, I weakly think the first alternative above is most useful.

Reply:
I definitely share the preference to express this function. I don't quite understand this; what would this look like for our `Pair` constructor? (I am excited to think this is actually possible!)

Reply:
```
val pair_val : 'z t -> (('z, 'a * 'b) Type_equal.t * 'a * 'b) option
```

Reply:
In general, all gadts without existential types can be turned into regular constructors with a Type_equal.t as extra payload. Deriving variants on such a regular constructor would do something, and that something is the type I wrote above for pair_val and earlier for `a_val`.

Reply:
I see; I tried to [write an implementation for this](https://sketch.sh/s/zaRSIGE8iRsQaWRhkb68KP/) but I am seeing type errors. Apologies if I have misunderstood this.

Reply:
Hm, sorry, I I'm pretty sure my advice just doesn't work when there are type variables.

Because `Pair` is equivalent to:

```
Pair of ('z, 'a * 'b) Type_equal.t * 'a t * ' b t
```

but the problem is that that type has existential variables in it, unlike the initial one. So we trade one problem for another problem.

Reply:
Ok; I guess the pragmatic thing to do is skip getters for now. This can always be added subsequently if a clear use case emerges for which we can always generate the functions.

Reply:
Agreed.